src/Compiler/Utilities/Caches.fs

```diff
@@ -42,180 +150,149 @@ type CachedEntity<'Key, 'Value> =
 
-    member _.GetInstanceStats() = [ "hit-ratio", ratio ] |> Map.ofList
+    // Track disposal state (0 = not disposed, 1 = disposed)
+    let mutable disposed = 0
 
-    static member val Meter = meter
+    let mutable deadKeysCount = 0
 
-    static member GetTotals(cacheId) =
-        observedCaches[cacheId].GetInstanceTotals()
+    // Keys with unreliable identity can prevent eviction, taking up space in the cache.
+    // In such case we rebuild the store to remove dead keys.
+    let rebuildStore () =
+        let newStore =
+            ConcurrentDictionary<'Key, CachedEntity<'Key, 'Value>>(Environment.ProcessorCount, capacity, options.Comparer)
 
-    static member GetStats(cacheId) =
-        observedCaches[cacheId].GetInstanceStats()
+        for entity in evictionQueue do
+            newStore.TryAdd(entity.Key, entity) |> ignore
 
-module Cache =
-    // During testing a lot of compilations are started in app domains and subprocesses.
-    // This is a reliable way to pass the override to all of them.
-    [<Literal>]
-    let private overrideVariable = "FSHARP_CACHE_OVERRIDE"
+        Interlocked.Exchange(&store, newStore) |> ignore
 
-    /// Use for testing purposes to reduce memory consumption in testhost and its subprocesses.
-    let OverrideCapacityForTesting () =
-        Environment.SetEnvironmentVariable(overrideVariable, "4096", EnvironmentVariableTarget.Process)
+    let processEvictionMessage msg =
+        match msg with
+        | EvictionQueueMessage.Add(entity: CachedEntity<_, _>, target) when isNull entity.Node.List ->
+            evictionQueue.AddLast(entity.Node)
+            // store has been rebuilt while this message was in the queue.
+            if store <> target then
+                store.TryAdd(entity.Key, entity) |> ignore
 
-    let applyOverride (capacity: int) =
-        match Int32.TryParse(Environment.GetEnvironmentVariable(overrideVariable)) with
-        | true, n when capacity > n -> n
-        | _ -> capacity
+            // Evict one immediately if necessary.
+            if evictionQueue.Count > capacity then
+                let first = nonNull evictionQueue.First
+                evictionQueue.Remove(first)
 
-[<Struct>]
-type EvictionQueueMessage<'Key, 'Value> =
-    | Add of CachedEntity<'Key, 'Value>
-    | Update of CachedEntity<'Key, 'Value>
+                match store.TryRemove(first.Value.Key) with
+                | true, _ ->
+                    CacheMetrics.Eviction tag
+                    evicted.Trigger()
+                | _ ->
+                    CacheMetrics.EvictionFail tag
+                    evictionFailed.Trigger()
+                    deadKeysCount <- deadKeysCount + 1
 
-[<Sealed; NoComparison; NoEquality>]
-[<DebuggerDisplay("{GetStats()}")>]
-type Cache<'Key, 'Value when 'Key: not null and 'Key: equality> internal (totalCapacity, headroom, name, listen) =
+                    if deadKeysCount > headroom / 2 then
+                        rebuildStore ()
```

Comment:
It would be great if we could tell this de-generate case happened.
Not yet sure how to communicate the message back :((

Reply:
(I have set autocomplete on - feel free to resolve this comment to get the PR in.)